### PR TITLE
Enables jailer to copy cache topology

### PR DESCRIPTION
**Pull request by xibz**
unfortunately due to differences in arm, arm does not setup a cache
topology. This commit fixes that copying the system's cache topology
into the jailed rootfs

Signed-off-by: xibz <impactbchang@gmail.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kzys** [comment]
Firecacker's jailer is doing something similar. Should it be owned by Go SDK?

https://github.com/firecracker-microvm/firecracker/blob/646ec96733d6aa72586e69dc9bcd0b38dab2df7d/src/jailer/src/env.rs#L282

**Comment by xibz** [comment]
@kzys the SDK has no knowledge of jailing. So no, it probably doesn't belong there imo

**Comment by kzys** [comment]
Oh sorry. This is for runc...

File: runtime/runc_jailer.go [hunk 1]
```diff
@@ -41,8 +41,19 @@ import (
...
 )
 
+var cacheTopologyPaths = []string{
+	"sys",
+	"devices",
+	"system",
+	"cpu",
+	"cpu0",
+	"cache",
+}
```
Comment (kzys) [inline runtime/runc_jailer.go:55]: - Does the directory has `cpu0` under `cpu0/cache`?
- What we copy is different from what Firecracker's jailer copies in https://github.com/firecracker-microvm/firecracker/pull/2334. Do you know the reason? (cc @dianpopa)

File: runtime/runc_jailer.go [hunk 2]
```diff
@@ -393,6 +408,25 @@ func (j *runcJailer) copyFileToJail(src, dst string, mode os.FileMode) error {
...
+// copyDirectoryToJail will copy all of the files and the parent directory to
+// the given destination folder, dst. This will also set the proper ownership
+// to the files and folders.
+func (j *runcJailer) copyDirectoryToJail(src, dst string) error {
+	output, err := exec.Command("cp", "-rf", src, dst).CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "failed to copy directory %q to %q: %s", src, dst, output)
+	}
+
+	filepath.Walk(dst, func(path string, info os.FileInfo, err error) error {
```
Comment (kzys) [inline runtime/runc_jailer.go:420]: I think we need to check the error from filepath.Walk.

File: runtime/runc_jailer.go [hunk 3]
```diff
@@ -640,3 +674,43 @@ func (j runcJailer) Stop(force bool) error {
...
+
+	// builds the cache topology path from the root directory
+	for _, p := range cacheTopologyPaths {
+		path = filepath.Join(path, p)
+		if err := mkdirAndChown(path, mode, j.Config.UID, j.Config.GID); err != nil {
+			return err
+		}
+	}
+
+	filepath.Walk(cacheTopologyPath, func(cachePath string, info os.FileInfo, err error) error {
```
Comment (kzys) [inline runtime/runc_jailer.go:693]: Check the error from `filepath.Walk`

**Review by kzys — CHANGES_REQUESTED** [review]

File: runtime/runc_jailer.go (hunk 1 above)
Comment (xibz) [reply runtime/runc_jailer.go:55]: Yea, I asked Diana and she told me today what files to copy. And yes, all these directories will exist.

File: runtime/runc_jailer.go (hunk 2 above)
Comment (xibz) [reply runtime/runc_jailer.go:420]: Will fix

**Comment by xibz** [comment]
@kzys Updated to copy the required files for the cache topology and the `filepath.Walk` error

File: runtime/runc_jailer.go [hunk 4]
```diff
@@ -640,3 +664,52 @@ func (j runcJailer) Stop(force bool) error {
...
+			return err
+		}
+		if !info.IsDir() {
+			return err
+		}
+
+		folder := filepath.Base(cachePath)
+		if !strings.HasPrefix(folder, cacheFolderPrefix) {
+			return err
+		}
```
Comment (kzys) [inline runtime/runc_jailer.go:694]: How about returning nil explicitly here?

**Review by kzys — APPROVED** [review]

File: runtime/runc_jailer.go (hunk 4 above)
Comment (xibz) [reply runtime/runc_jailer.go]: Cool, will fix this. Forgot that I returned if `err` was not `nil` at the top of the closure.

**Pull request merged**